[DI] Allow dumping the container in one file instead of many files

lyrixx opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 4.4
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | -
| Tests pass?   | yes    <!-- please add some, will be required by reviewers -->
| Fixed tickets | -
| License       | MIT
| Doc PR        | -

Replaced by #32581

lyrixx committed:
[Kernel+DIC] Fixed as_files behavior

Many issue about this:

* Kernel does not handle when the dumper returns a string
* Kernel does not handle when the compiled Container file does not
return a new instance of itself
* Kernel try to old load and clean old container, but if the container
does not change its name, it tries to load 2 time the same file, and so
the same class

This patch fixed all theses case. Now we can use with ease a multifiles
container or a simple unique file.

stof commented on src/Symfony/Component/HttpKernel/Kernel.php:
```diff
@@ -835,7 +836,7 @@ protected function getContainerBuilder()
      * @param string           $baseClass The name of the container's base class
      */
-    protected function dumpContainer(ConfigCache $cache, ContainerBuilder $container, $class, $baseClass)
+    protected function dumpContainer(ConfigCache $cache, ContainerBuilder $container, $class, $baseClass, bool $asFiles = true)
```
is this ever passed as `false` ?

lyrixx replied:
if you change the $asFile that are hardcoded... yes :) I just fixed the other side of an hardcoded path :joy:

nicolas-grekas replied:
Instead of an argument, this should be a parameter to me.
That would fix the BC break and make this easier to configure I believe.

lyrixx replied:
a DIC parameter you mean?

nicolas-grekas replied:
yes

lyrixx replied:
Fixed

nicolas-grekas commented on src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php:
```diff
@@ -269,6 +269,8 @@ public function dump(array $options = [])
                 $code .= $c;
             }
+
+            $code .= sprintf("\nreturn new %s();\n", $options['class']);
```
this should be removed (moved to the dumpContainer method)

lyrixx replied:
why. like that, it's consistent with the "asFiles=true" mode!

nicolas-grekas replied:
because there are ppl out there that use the component standalone an don't want a container to be created and destroyed immediately :)

lyrixx replied:
Nice! So this is a bug fix !

lyrixx replied:
Fixed

nicolas-grekas reviewed (commented):
That's for 4.4, it's a new feature :)

lyrixx reviewed (commented):
Ok, I have reworked this PR. there was a bug (did not worked in env test), now it's :+1: 
Bonus: Now it's only affect the kernel, not the DIC \o/

lyrixx commented on src/Symfony/Component/HttpKernel/Kernel.php:
```diff
@@ -572,18 +572,25 @@ protected function getContainerBaseClass()
             try {
                 if (file_exists($cache->getPath()) && \is_object($this->container = include $cache->getPath())) {
                     $this->container->set('kernel', $this);
-                    $oldContainer = $this->container;
```
This var was useless, because at the new line we have `$fresh=true` and later `if ($fresh} { return; }`

nicolas-grekas replied:
unless the line above throws an exception, which is why this code is how it is

lyrixx replied:
:+1:

lyrixx replied:
I don't understand: 
* If the line above throw an exception, then we can not execute this line.
* If the line above does not throw exception, so we execute this line. But this line can not fail. So we execute the next line: `$fresh=true` and later `if ($fresh} { return; }`. So to me this is dead code.

lyrixx commented:
The diff is easier to read thank to [`?w=1`](https://github.com/symfony/symfony/pull/32119/files?w=1)

nicolas-grekas commented on src/Symfony/Component/HttpKernel/Kernel.php:
```diff
@@ -844,27 +863,33 @@ protected function dumpContainer(ConfigCache $cache, ContainerBuilder $container
             $dumper->setProxyDumper(new ProxyDumper());
         }
 
+        $asFiles = $container->hasParameter('container.as_files') ? $container->getParameter('container.as_files') : true;
```
container.dump_as_files?
or
container.dumper.as_files?

lyrixx replied:
Actually, I wanted to stick the current format

https://github.com/symfony/symfony/blob/64eaf7e3a63e6586f18582ea4ab8e5f334560525/src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php#L298-L300

Should I change? If so, I prefer `container.dumper.as_files`

nicolas-grekas replied:
Double checking, `container.dumper.*` is for configuring the dumper,
while this one configures how the kernel uses the dumper...
So I'm for `container.dump_as_files`, which is the most explicit.

lyrixx replied:
okay, I will change this

lyrixx commented:
Are you OK with this one?

nicolas-grekas commented on src/Symfony/Component/HttpKernel/Kernel.php:
```diff
@@ -572,18 +572,25 @@ protected function getContainerBaseClass()
         $cache = new ConfigCache($cacheDir.'/'.$class.'.php', $this->debug);
         $oldContainer = null;
         if ($fresh = $cache->isFresh()) {
+            if (!$asFiles) {
```
this cannot happen

lyrixx replied:
:) If I change the value line 575, it will happen.
But OK, I will add something to be able to change it in a clean manier. So 4.4, you win!
But I don't know how to do that with the DIC. I think the only solution is to add a new paratemer (the DIC is not build at this time)

nicolas-grekas commented on src/Symfony/Component/HttpKernel/Kernel.php:
```diff
@@ -572,18 +572,25 @@ protected function getContainerBaseClass()
         $oldContainer = null;
         if ($fresh = $cache->isFresh()) {
+            if (!$asFiles) {
+                require_once $cache->getPath();
```
moving this outside of the try/catch+require_cone where it was before means this will fail badly when the file is missing (will happen on race conditions)

lyrixx replied:
This code was present in symfony 2.X + there is not try/catch block here. So I don't get what you mean

lyrixx replied:
I moved that in the catch block

nicolas-grekas commented on src/Symfony/Component/HttpKernel/Kernel.php:
```diff
@@ -634,6 +641,13 @@ protected function initializeContainer()
         try {
             $container = null;
             $container = $this->buildContainer();
+            if ($asFiles) {
```
always true here - but why is this needed? it's not right now

lyrixx replied:
Same answer, I will add a new param

nicolas-grekas commented on src/Symfony/Component/HttpKernel/Kernel.php:
```diff
@@ -634,6 +641,13 @@ protected function initializeContainer()
+                $container->getParameterBag()->add([
+                    'container.build_id' => 'static',
+                    'container.build_hash' => 'static',
+                    'container.as_files' => true,
```
why is this needed? this should already be dumped as part of the regular container *when the param is defined*

lyrixx replied:
This is needed to avoid adding an extra param to the method `dumpContainer()`

lyrixx replied:
Fixed

nicolas-grekas commented on src/Symfony/Component/HttpKernel/Kernel.php:
```diff
@@ -657,7 +671,12 @@ protected function initializeContainer()
 
         $this->dumpContainer($cache, $container, $class, $this->getContainerBaseClass());
-        $this->container = require $cache->getPath();
+        if ($asFiles) {
```
always true also, or I'm missing something

nicolas-grekas reviewed (commented):
still for 4.4 to me, sorry about that :)

lyrixx committed:
Simplified code

lyrixx commented:
OK, I have added a test, and addressed your comments.
I also added an extra commit where I simplified the code. (about fresh / oldContainer) May be I miss something, but I think you are wrong. Anyway, It's an extra commit, so I would remove it easily! 
So it could be better to review the first commit first

lyrixx commented:
I rebased this PR. It's now ready and green :green_heart:

nicolas-grekas commented:
I pushed a completely different implementation on your branch: instead of patching the Kernel (I kept your cleanup), this now patches the PhpDumper. This allows keeping the anonymous class proxy, which is desired to have safe cache clears.

nicolas-grekas commented:
Replaced by #32581
Thanks @lyrixx for raising the issue and giving it a try.

lyrixx commented:
@nicolas-grekas  What about the cleaning ?
